pkg/osd_cluster_upgrader/upgrader.go

```diff
@@ -567,3 +580,35 @@ func hasUpgradeCommenced(c client.Client, uc *upgradev1alpha1.UpgradeConfig) (bo
 
 	return true, nil
 }
+
+// nodeDrained checks if the nodes are being drained successfully during the upgrade
+func nodeDrained(c client.Client, uc *upgradev1alpha1.UpgradeConfig, logger logr.Logger, metrics metrics.Metrics) (bool, error) {
+	nodeList := &corev1.NodeList{}
+	errNode := c.List(context.TODO(), nodeList)
+	if errNode != nil {
+		return false, errNode
+	}
+
+	pdbAlertName := "PodDisruptionBudgetAtLimit"
+	pdbAlerts, errMetric := metrics.Query(fmt.Sprintf("ALERTS{alertname=\"%s\"}", pdbAlertName))
```

Comment:
So if we only allow Critical alerts to be raised during an upgrade. This will never be detected as it `severity:warning` as per: https://github.com/openshift/cluster-kube-controller-manager-operator/blob/master/manifests/0000_90_kube-controller-manager-operator_05_alert-pdb.yaml#L12

We ignore `warnings` in our own silences: https://github.com/openshift/managed-upgrade-operator/blob/master/pkg/maintenance/alertmanagerMaintenance.go#L185

Reply:
The silence was created in Alertmanager, where this call is calling the prometheus.

The prometheus will always fire the alert with specified severity, the alertmanager will decide if alert with integrated tools like PagerDuty.

So it should not be a problem for us.

Reply:
I was thinking about this too and I think its better to check for `PodDisruptionBudget` objects directly instead of querying prometheus. This removes prometheus and keeps it Kubernetes.
Looking at the alert it checks `status_expected_pods == status_desired_healthy`

Reply:
@sedroche Moved the pdb detection out of prometheus